src/result-async.ts

```diff
@@ -0,0 +1,123 @@
+import { Result, Ok, Err } from './'
+import { isPromise } from './'
+
+export class ResultAsync<T, E> {
+  private _promise: Promise<Result<T, E>>
+
+  constructor(res: Promise<Result<T, E>>) {
+    this._promise = res
+  }
+
+  static fromPromise<T, E>(promise: Promise<T>): ResultAsync<T, E> {
+    return new ResultAsync(promise.then((value: T) => new Ok<T, E>(value)))
+  }
+
+  map<A>(_f: (t: T) => A | Promise<A>): ResultAsync<A, E> {
+    const newPromise: Promise<Result<A, E>> = this._promise.then((res: Result<T, E>) => {
+      if (res.isErr()) {
+        return new Err<A, E>(res.error)
+      }
+
+      const newValue = _f(res.value)
+
+      if (isPromise(newValue)) {
+        return newValue.then(value => new Ok<A, E>(value))
+      }
+
+      return new Ok<A, E>(newValue)
+    })
+
+    return new ResultAsync(newPromise)
+  }
+
+  mapErr<U>(_f: (e: E) => U | Promise<U>): ResultAsync<T, U> {
+    const newPromise: Promise<Result<T, U>> = this._promise.then((res: Result<T, E>) => {
+      if (res.isOk()) {
+        return new Ok<T, U>(res.value)
+      }
+
+      const newError = _f(res.error)
+
+      if (isPromise(newError)) {
+        return newError.then(error => new Err<T, U>(error))
+      }
+
+      return new Err<T, U>(newError)
+    })
+
+    return new ResultAsync(newPromise)
+  }
+
+  andThen<U>(_f: (t: T) => Result<U, E> | ResultAsync<U, E>): ResultAsync<U, E> {
+    const newPromise = this._promise.then((res: Result<T, E>) => {
+      if (res.isErr()) {
+        return new Err<U, E>(res.error)
+      }
+
+      const newValue = _f(res.value)
+
+      if (newValue instanceof ResultAsync) {
+        return newValue._promise
+      }
+
+      return newValue
+    })
+
+    return new ResultAsync(newPromise)
+  }
+
+  match = <A>(ok: (t: T) => A, _err: (e: E) => A): Promise<A> => {
+    return this._promise.then(res => res.match(ok, _err))
+  }
+
+  // Makes ResultAsync awaitable
+  then(_successCallback: (res: Result<T, E>) => void) {
```

Comment:
Cool! I learned something new :) By the way, does this method need to return anything? Seems counter-intuitive that this method doesn't return a value. 

In my head I am thinking, "in normal promises, I can chain `.then` as many times as I want ... so wouldn't that imply that a `.then` method returns a `Promise`?"

Reply:
You are absolutely correct.

I added a test (which of course didn't pass with the previous version) and fixed the implementation.